src/services/ansiblePlaybook.ts

```diff
@@ -0,0 +1,179 @@
+import { withInterpreter } from '../utils/misc';
+import { WorkspaceFolderContext } from './workspaceManager';
+const exec = promisify(child_process.exec);
+
+/**
+ * Acts as an interface to ansible-playbook command.
+ */
+export class AnsiblePlaybook {
+  private connection: Connection;
+  private context: WorkspaceFolderContext;
+  private useProgressTracker = false;
+
+  constructor(connection: Connection, context: WorkspaceFolderContext) {
+    this.connection = connection;
+    this.context = context;
+    this.useProgressTracker =
+      !!context.clientCapabilities.window?.workDoneProgress;
+  }
+
+  /**
+   * Acts as an interface to ansible-playbook <file> --syntax-check command and a cache of its output.
+   * ansible syntax-check may provide diagnostics for more than just the file for which
+   * it was triggered, and this is reflected in the implementation.
+   *
+   * Perform ansible syntax-check for the given document.
+   */
+  public async doValidate(
+    textDocument: TextDocument
+  ): Promise<Map<string, Diagnostic[]>> {
+    const docPath = new URL(textDocument.uri).pathname;
+    let diagnostics: Map<string, Diagnostic[]> = new Map();
+    let progressTracker;
+    if (this.useProgressTracker) {
+      progressTracker = await this.connection.window.createWorkDoneProgress();
+    }
+
+    const workingDirectory = new URL(this.context.workspaceFolder.uri).pathname;
+
+    try {
+      const settings = await this.context.documentSettings.get(
+        textDocument.uri
+      );
+
+      if (progressTracker) {
+        progressTracker.begin(
+          'ansible syntax-check',
+          undefined,
+          'Processing files...'
+        );
+      }
+
+      const [command, env] = withInterpreter(
+        `${settings.ansible.path}-playbook`,
+        `${docPath} --syntax-check`,
+        settings.python.interpreterPath,
+        settings.python.activationScript
+      );
+
+      await exec(command, {
+        encoding: 'utf-8',
+        cwd: workingDirectory,
+        env: env,
+      });
+    } catch (error) {
+      if (error instanceof Error) {
+        const execError = error as ExecException & {
+          // according to the docs, these are always available
+          stdout: string;
+          stderr: string;
+        };
+
+        // This is the regex to extract the filename, line and column number from the strerr produced by syntax-check command
+        const ansibleSyntaxCheckRegex =
+          /The error appears to be in '(?<filename>.*)': line (?<line>\d+), column (?<column>\d+)/;
+
+        const result = ansibleSyntaxCheckRegex.exec(execError.stderr);
+
+        diagnostics = result
+          ? this.processReport(
+              execError.message,
+              result.groups.filename,
+              parseInt(result.groups.line),
+              parseInt(result.groups.column)
+            )
+          : this.processReport(execError.message, docPath, 1, 1);
```

Comment:
How about
```suggestion
        diagnostics = this.processReport(
          execError.message,
          result ? result.groups.filename : docPath,
          result ? parseInt(result.groups.line) : 1,
          result ? parseInt(result.groups.column) : 1,
        );
```

Reply:
I kept it that way so that I have a clear distinction for different cases (by not using the ternary operator at multiple keys and keeping the returned objects separate).